### PR TITLE
Scope hoisting destructuring

### LINKED ISSUE #2696: Experimental scope hoisting: importing Web3 causes error
# 🐛 bug report

I'm trying out the `--experimental-scope-hoisting` flag, and it doesn't play nice with Web3. Can't say I'm surprised, Web3 is quite a large library and still in a beta stage. When I saw it wasn't working, I starting cutting out code until I was only importing Web3 and nothing else, which still resulted in an error during packaging with `Property name expected type of string but got null`.

I think I'll simply remove Web3 as an NPM dependency and load it temporarily from a CDN. It would be awesome to see them play nicely together at some point :)

## 🎛 Configuration (.babelrc, package.json, cli command)

`web3@1.0.0-beta.46` in package.json

.babelrc: just the default provided by ParcelJS. I'm not using a custom .babelrc.

cli: `parcel build src/web/index.html --experimental-scope-hoisting"`

## 🤔 Expected Behavior

Expecting a little tree-shaking ;)

## 😯 Current Behavior

```
Property name expected type of string but got null
    at validate (/Users/codyredmond/Documents/Work/everyday/node_modules/@babel/types/lib/definitions/utils.js:161:13)
    at Object.validate (/Users/codyredmond/Documents/Work/everyday/node_modules/@babel/types/lib/definitions/utils.js:172:7)
    at validate (/Users/codyredmond/Documents/Work/everyday/node_modules/@babel/types/lib/validators/validate.js:17:9)
    at builder (/Users/codyredmond/Documents/Work/everyday/node_modules/@babel/types/lib/builders/builder.js:46:27)
    at Object.Identifier (/Users/codyredmond/Documents/Work/everyday/node_modules/@babel/types/lib/builders/generated/index.js:321:31)
    at JSConcatPackager.wrapModule (/Users/codyredmond/Documents/Work/everyday/node_modules/parcel-bundler/src/packagers/JSConcatPackager.js:292:21)
    at JSConcatPackager.addDeps (/Users/codyredmond/Documents/Work/everyday/node_modules/parcel-bundler/src/packagers/JSConcatPackager.js:239:25)
    at JSConcatPackager.addDeps (/Users/codyredmond/Documents/Work/everyday/node_modules/parcel-bundler/src/packagers/JSConcatPackager.js:223:27)
    at JSConcatPackager.addDeps (/Users/codyredmond/Documents/Work/everyday/node_modules/parcel-bundler/src/packagers/JSConcatPackager.js:223:27)
    at JSConcatPackager.addDeps (/Users/codyredmond/Documents/Work/everyday/node_modules/parcel-bundler/src/packagers/JSConcatPackager.js:223:27)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! everyday@0.1.0 build: `parcel build src/web/index.html --experimental-scope-hoisting`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the everyday@0.1.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/codyredmond/.npm/_logs/2019-02-28T02_57_40_473Z-debug.log
```

## 💁 Possible Solution

I'll be loading web3 temporarily from a CDN. I've been reading over at web3 that they aren't too happy about the large sizes and inaccurate typing, so at least we know they're aware of it. I'm also curious about https://github.com/xf00f/web3x, but with the speed of development on web3, I'm not sure about how it can claim parity. I'll be looking into it.

## 🔦 Context

I just would love to keep using ParcelJS for building out ethereum apps. What an amazing bundler you have :D

## 💻 Code Sample

This is a work in progress, but feel free trying to create a production build with experimental hoisting turned on using this feature branch: https://github.com/codydjango/everyday/tree/feature/experimental-scope-hoisting

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 1.11.0
| Node             | v11.6.0
| npm/Yarn         | 6.7.0
| Operating System | osx 10.14.3

### PR DESCRIPTION
**Pull request by mischnic**
# ↪️ Pull Request

Part of #2696

## 💻 Examples

If scope hoisting needs to wrap a module and hoist all variables declarations in the outermost scope, a untranspiled array and object destructuring would trip up Parcel.

## ✔️ PR Todo

- [x] Added/updated unit tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

### CODE REVIEW

File: packages/core/parcel-bundler/src/packagers/JSConcatPackager.js [hunk 1]
```diff
@@ -283,17 +283,50 @@ class JSConcatPackager extends Packager {
...
-          decls.push(t.variableDeclarator(decl.id));
-          if (decl.init) {
-            body.push(
-              t.expressionStatement(
-                t.assignmentExpression(
-                  '=',
-                  t.identifier(decl.id.name),
-                  decl.init
+          if (t.isObjectPattern(decl.id)) {
+            for (let prop of decl.id.properties) {
```
Comment (devongovett) [inline packages/core/parcel-bundler/src/packagers/JSConcatPackager.js]: I believe ObjectPattern and ArrayPattern work as the left hand side of an AssignmentExpression, in addition to a VariableDeclarator. e.g. `var {x} = y` -> `({x} = y)`. That would be smaller in the output.

File: packages/core/parcel-bundler/src/packagers/JSConcatPackager.js [hunk 2]
```diff
@@ -283,17 +283,50 @@ class JSConcatPackager extends Packager {
...
+                      '=',
+                      t.identifier(prop.value.name),
+                      t.memberExpression(decl.init, prop.key)
+                    )
+                  )
+                );
+              }
+            }
+          } else if (t.isArrayPattern(decl.id)) {
+            for (let i = 0; i < decl.id.elements.length; i++) {
```
Comment (devongovett) [inline packages/core/parcel-bundler/src/packagers/JSConcatPackager.js]: I think you can do the same here: `var [x] = y` -> `[x] = y`

File: packages/core/parcel-bundler/src/packagers/JSConcatPackager.js (hunk 1 above)
Comment (mischnic) [reply packages/core/parcel-bundler/src/packagers/JSConcatPackager.js]: Done, and `getBindingIdentifiers` did indeed remove a whole if case

**Comment by mischnic** [comment]
> Awesome! Could you port this to the v2 branch as well?

Yes, had already planned doing that.

**Pull request merged**